src/End.sol

```diff
@@ -0,0 +1,335 @@
+    event Flow(bytes32 indexed ilk);
+    event Pack(address indexed usr, uint256 wad);
+    event Cash(bytes32 indexed ilk, address indexed usr, uint256 wad);
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "End/not-authorized");
+        _;
+    }
+
+    // --- Init ---
+    constructor(address vat_) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        vat = VatLike(vat_);
+        live = 1;
+    }
+
+    // --- Math ---
+    uint256 constant WAD = 10 ** 18;
+    uint256 constant RAY = 10 ** 27;
+    function min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        return x <= y ? x : y;
+    }
+    function rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x * y / RAY;
+    }
+    function wdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x * WAD / y;
+    }
+
+    // --- Administration ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        require(live == 1, "End/not-live");
+        if (what == "vow") vow = data;
+        else if (what == "pot") pot = PotLike(data);
+        else if (what == "spot") spot = SpotLike(data);
+        else if (what == "cure") cure = CureLike(data);
+        else if (what == "claim") claim = ClaimLike(data);
+        else revert("End/file-unrecognized-param");
+        emit File(what, data);
+    }
+    function file(bytes32 what, uint256 data) external auth {
+        require(live == 1, "End/not-live");
+        if (what == "wait") wait = data;
+        else revert("End/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Settlement ---
+    function cage() external auth {
+        require(live == 1, "End/not-live");
+        live = 0;
+        when = block.timestamp;
+        vat.cage();
+        spot.cage();
+        pot.cage();
+        cure.cage();
+        emit Cage();
+    }
+
+    function cage(bytes32 ilk) external {
+        require(live == 0, "End/still-live");
+        require(tag[ilk] == 0, "End/tag-ilk-already-defined");
+        (Art[ilk],,,,) = vat.ilks(ilk);
+        (PipLike pip,) = spot.ilks(ilk);
+        // par is a ray, pip returns a wad
+        tag[ilk] = wdiv(spot.par(), uint256(pip.read()));
+        emit Cage(ilk);
+    }
+
+    function skim(bytes32 ilk, address urn) external {
+        require(tag[ilk] != 0, "End/tag-ilk-not-defined");
+        (, uint256 rate,,,) = vat.ilks(ilk);
+        (uint256 ink, uint256 art) = vat.urns(ilk, urn);
+
+        uint256 owe = rmul(rmul(art, rate), tag[ilk]);
+        uint256 wad = min(ink, owe);
+        gap[ilk] = gap[ilk] + (owe - wad);
+
+        require(wad <= 2**255 && art <= 2**255, "End/overflow");
+        vat.grab(ilk, urn, address(this), vow, -int256(wad), -int256(art));
```

Comment:
In Solidity >= 0.8 `-int256(wad)` will revert for `wad == 2**255`, because `int256(2**255)` is actually `-2**255`, inversion of which is not an `int256`. Check for `wad` should be: `wad <= type(int256).max`